### ISSUE TITLE
0.55 regression around (?) setitem(readonly array(int64, 1d, C), Literal[int](0), Literal[int](0)))

### ISSUE THREAD

**Issue by slayoo** [numba/numba#7738]
The following code works OK with 0.54 (as well as with NUMBA_DISABLE_JIT=1) but fails with 0.55:
```python
import numpy as np
import numba

@numba.njit()
def fill_halos():
    return 0

@numba.njit()
def sset(arr, k, value):
    arr[k] = value

@numba.njit()
def boundary_cond_scalar(psi, fun_inner):
    sset(psi, 0, fun_inner())

@numba.njit()
def apply_scalar(arg5s_data, arg5s_bc_i):
    boundary_cond_scalar(arg5s_data, arg5s_bc_i)

null_scalarfield = (np.full((5,), 44),)

@numba.njit()
def apply():
    return apply_scalar(null_scalarfield[0], fill_halos)

def test_formulae_upwind():
    apply()
```

errors are:

```
E           Failed in nopython mode pipeline (step: nopython frontend)
E           Failed in nopython mode pipeline (step: nopython frontend)
E           Failed in nopython mode pipeline (step: nopython frontend)
E           No implementation of function Function(<built-in function setitem>) found for signature:
E            
E            >>> setitem(readonly array(int64, 1d, C), Literal[int](0), Literal[int](0))
E            
E           There are 16 candidate implementations:
E             - Of which 14 did not match due to:
E             Overload of function 'setitem': File: <numerous>: Line N/A.
E               With argument(s): '(readonly array(int64, 1d, C), int64, int64)':
E              No match.
E             - Of which 2 did not match due to:
E             Overload in function 'SetItemBuffer.generic': File: numba/core/typing/arraydecl.py: Line 176.
E               With argument(s): '(readonly array(int64, 1d, C), int64, int64)':
E              Rejected as the implementation raised a specific error:
E                NumbaTypeError: Cannot modify readonly array of type: readonly array(int64, 1d, C)
E             raised from /home/slayoo/devel/venvs/python3.8/lib/python3.8/site-packages/numba/core/typing/arraydecl.py:183
E           
E           During: typing of setitem at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (10)
E           
E           File "test_numba_0_55.py", line 10:
E           def sset(arr, k, value):
E               arr[k] = value
E               ^
E           
E           During: resolving callee type: type(CPUDispatcher(<function sset at 0x7f865d1d10d0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (14)
E           
E           During: resolving callee type: type(CPUDispatcher(<function sset at 0x7f865d1d10d0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (14)
E           
E           
E           File "test_numba_0_55.py", line 14:
E           def boundary_cond_scalar(psi, fun_inner):
E               sset(psi, 0, fun_inner())
E               ^
E           
E           During: resolving callee type: type(CPUDispatcher(<function boundary_cond_scalar at 0x7f865d1d1af0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (18)
E           
E           During: resolving callee type: type(CPUDispatcher(<function boundary_cond_scalar at 0x7f865d1d1af0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (18)
E           
E           
E           File "test_numba_0_55.py", line 18:
E           def apply_scalar(arg5s_data, arg5s_bc_i):
E               boundary_cond_scalar(arg5s_data, arg5s_bc_i)
E               ^
E           
E           During: resolving callee type: type(CPUDispatcher(<function apply_scalar at 0x7f865d1d1ca0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (24)
E           
E           During: resolving callee type: type(CPUDispatcher(<function apply_scalar at 0x7f865d1d1ca0>))
E           During: typing of call at /home/slayoo/devel/PyMPDATA/tests/unit_tests/test_numba_0_55.py (24)
E           
E           
E           File "test_numba_0_55.py", line 24:
E           def apply():
E               return apply_scalar(null_scalarfield[0], fill_halos)
E               ^

```

**Comment by stuartarchibald** [comment]
Thanks for the report. Reducing the example code to:
```python
import numpy as np
import numba

global_tuple = (np.zeros((1,)),)

@numba.njit
def foo(x):
    global_tuple[0][x] = 0

foo(0)
```
I think this is relying on a bug (https://github.com/numba/numba/issues/7120) that was fixed in 0.55 (patch https://github.com/numba/numba/pull/7307). Essentially, the code is trying to mutate a global and Numba 0.55 correctly prevents that from occurring. I've opened https://github.com/numba/numba/issues/7723 for discussion over how to make it more clear that this is the issue, please do comment on/subscribe to that issue if you are interested. Thanks.

**Comment by slayoo** [comment]
Thank you! Closing this one then.
